File: Week 26/python/min_cut.py
"""
WEEK 26 - PYTHON ADVANCED TOPICS
Topic: Minimum s-t Cut via Max-Flow / Min-Cut Theorem
File: min_cut.py

CONCEPT:
    Given a directed flow network with source `s` and sink `t`, an s-t cut is
    a partition (S, T) of the vertices with s in S and t in T. The cut's
    capacity is the sum of capacities of edges from S to T (forward only).
    The Max-Flow / Min-Cut theorem states:

        max flow value = min cut capacity.

    To extract the cut after running any max-flow algorithm, perform a BFS
    on the *residual* graph from s; vertices reachable belong to S, the rest
    to T. Cut edges are precisely the original edges (u,v) with u in S and
    v in T whose residual is now zero (i.e. saturated).

KEY POINTS:
    - Works on top of any max-flow algorithm (we use Edmonds-Karp here).
    - Distinguish ORIGINAL edges from residual reverse edges — only original
      forward edges count toward the cut capacity.
    - Useful for: image segmentation, project selection, bipartite vertex
      cover (Konig's theorem), reliability analysis.

ALGORITHM / APPROACH:
    1. Run any max-flow algorithm. Now residual graph reflects flows.
    2. BFS/DFS from `s` in residual graph treating edges with cap > 0 as
       traversable. Mark visited set S.
    3. For every original edge (u,v) with u in S and v not in S: it is a
       cut edge with original capacity contributing to min-cut value.

PYTHON-SPECIFIC NOTES:
    - We tag each edge with a flag `is_original` so we can later distinguish
      forward edges from their auto-generated reverses.
    - Returning a list[tuple[int,int,int]] of (u, v, capacity) feels Pythonic.

DRY RUN / EXAMPLE:
    Same CLRS graph (max flow 23). After saturation, BFS from 0 in residual:
        Reachable set S = {0, 2, 4} (try and confirm with the trace).
        Cut edges from S to T = {1,3,5}: (0,1)=16, (4,3)=7 ... etc.
        Sum of original capacities = 23 = max flow.

COMPLEXITY:
    Same as the underlying max-flow algorithm + O(V+E) for the cut extraction.
"""

from __future__ import annotations

from collections import deque
from typing import List, Tuple


class MinCut:
    def __init__(self, n: int) -> None:
        self.n = n
        # Edge: [to, cap, rev_index, is_original_forward, original_cap]
        self.graph: List[List[List]] = [[] for _ in range(n)]

    def add_edge(self, u: int, v: int, cap: int) -> None:
        self.graph[u].append([v, cap, len(self.graph[v]), True, cap])
        self.graph[v].append([u, 0, len(self.graph[u]) - 1, False, 0])

    # --- Edmonds-Karp embedded for self-containment ----------------------
    def _bfs(self, s: int, t: int):
        parent = [None] * self.n
        parent[s] = (-1, -1)
        q = deque([s])
        while q:
            u = q.popleft()
            for idx, edge in enumerate(self.graph[u]):
                v, cap = edge[0], edge[1]
                if parent[v] is None and cap > 0:
                    parent[v] = (u, idx)
                    if v == t:
                        return parent
                    q.append(v)
        return None

    def max_flow(self, s: int, t: int) -> int:
        flow = 0
        while True:
            parent = self._bfs(s, t)
            if parent is None or parent[t] is None:
                return flow
            bottleneck = float("inf")
            v = t
            while v != s:
                u, eidx = parent[v]
                bottleneck = min(bottleneck, self.graph[u][eidx][1])
                v = u
            v = t
            while v != s:
                u, eidx = parent[v]
                edge = self.graph[u][eidx]
                edge[1] -= bottleneck
                self.graph[v][edge[2]][1] += bottleneck
                v = u
            flow += int(bottleneck)

    # --- Cut extraction --------------------------------------------------
    def reachable_from_source(self, s: int) -> List[bool]:
        visited = [False] * self.n
        visited[s] = True
        q = deque([s])
        while q:
            u = q.popleft()
            for v, cap, _rev, _orig, _ocap in self.graph[u]:
                if cap > 0 and not visited[v]:
                    visited[v] = True
                    q.append(v)
        return visited

    def min_cut_edges(self, s: int, t: int) -> Tuple[int, List[Tuple[int, int, int]]]:
        flow = self.max_flow(s, t)
        side = self.reachable_from_source(s)
        cut: List[Tuple[int, int, int]] = []
        for u in range(self.n):
            if not side[u]:
                continue
            for edge in self.graph[u]:
                v, _cap, _rev, is_orig, ocap = edge
                if is_orig and not side[v]:
                    cut.append((u, v, ocap))
        return flow, cut


def _demo() -> None:
    g = MinCut(6)
    edges = [
        (0, 1, 16), (0, 2, 13),
        (1, 2, 4),  (1, 3, 12),
        (2, 1, 10), (2, 4, 14),
        (3, 2, 9),  (3, 5, 20),
        (4, 3, 7),  (4, 5, 4),
    ]
    for u, v, c in edges:
        g.add_edge(u, v, c)
    flow, cut = g.min_cut_edges(0, 5)
    print(f"Max flow / Min cut value: {flow}")
    print("Cut edges (S -> T):")
    for u, v, c in cut:
        print(f"  ({u} -> {v}) capacity {c}")


if __name__ == "__main__":
    _demo()


# NOTES
# -----
# Differences from Java:
#   * Java's network_flow.java focuses on max-flow and matching; min-cut
#     extraction is implicit. Here we add an explicit `min_cut_edges` helper.
#   * The tuple (u, v, original_capacity) is a more Pythonic return type
#     than a custom Edge class.
#   * We attach the original capacity to each edge so that after max-flow
#     mutates residuals we can still report the *capacity* of cut edges.

File: Week 26/java/min_cut.java
/*
 * WEEK 26 - JAVA ADVANCED TOPICS
 * Topic: Minimum s-t Cut via Max-Flow / Min-Cut Theorem
 * File: min_cut.java
 *
 * CONCEPT:
 *     Given a directed flow network with source s and sink t, an s-t cut is
 *     a partition (S, T) of the vertices with s in S and t in T. The cut's
 *     capacity is the sum of capacities of edges from S to T (forward only).
 *     The Max-Flow / Min-Cut theorem states:
 *
 *         max flow value = min cut capacity.
 *
 *     To extract the cut after running any max-flow algorithm, BFS the
 *     residual graph from s; vertices reachable form S, the rest form T.
 *     Cut edges are the *original* forward edges (u,v) with u in S and v in
 *     T whose residual is now zero (saturated).
 *
 * KEY POINTS:
 *     - Works on top of any max-flow algorithm (we embed Edmonds-Karp).
 *     - Distinguish ORIGINAL edges from residual reverse edges; only
 *       originals count in the cut capacity.
 *     - Applications: image segmentation, project selection, bipartite
 *       vertex cover (Konig's theorem), reliability analysis.
 *
 * ALGORITHM / APPROACH:
 *     1. Run any max-flow algorithm. Residual now reflects flows.
 *     2. BFS from s in residual graph treating edges with cap > 0 as live.
 *        Mark reachable set S.
 *     3. For every original edge (u,v) with u in S and v not in S: it is a
 *        cut edge contributing its ORIGINAL capacity to the min-cut value.
 *
 * DRY RUN / EXAMPLE:
 *     Same CLRS graph (max flow 23). After saturation, BFS from 0 in
 *     residual reaches S = {0, 2, 4}. Cut edges (S -> T):
 *         (0,1) cap 16, (4,3) cap 7. Sum = 23 == max flow.
 *
 * COMPLEXITY:
 *     Same as the underlying max-flow algorithm + O(V+E) for cut extraction.
 */

// snake_case filename is fine because class MinCut is package-private.

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Deque;
import java.util.List;

class MinCut {

    static class Edge {
        int to, cap, rev;
        boolean original;
        int originalCap;
        Edge(int to, int cap, int rev, boolean original, int originalCap) {
            this.to = to; this.cap = cap; this.rev = rev;
            this.original = original; this.originalCap = originalCap;
        }
    }

    static class CutEdge {
        final int u, v, capacity;
        CutEdge(int u, int v, int capacity) { this.u = u; this.v = v; this.capacity = capacity; }
        @Override public String toString() { return "(" + u + " -> " + v + ") cap " + capacity; }
    }

    private final int n;
    private final List<List<Edge>> graph;

    MinCut(int n) {
        this.n = n;
        this.graph = new ArrayList<>();
        for (int i = 0; i < n; i++) graph.add(new ArrayList<>());
    }

    void addEdge(int u, int v, int cap) {
        graph.get(u).add(new Edge(v, cap, graph.get(v).size(), true, cap));
        graph.get(v).add(new Edge(u, 0, graph.get(u).size() - 1, false, 0));
    }

    // --- Embedded Edmonds-Karp for self-containment ---
    int maxFlow(int s, int t) {
        int flow = 0;
        while (true) {
            int[] parent = new int[n];
            int[] parentEdge = new int[n];
            Arrays.fill(parent, -1);
            parent[s] = s;
            Deque<Integer> queue = new ArrayDeque<>();
            queue.offer(s);
            while (!queue.isEmpty() && parent[t] == -1) {
                int u = queue.poll();
                List<Edge> adj = graph.get(u);
                for (int i = 0; i < adj.size(); i++) {
                    Edge e = adj.get(i);
                    if (parent[e.to] == -1 && e.cap > 0) {
                        parent[e.to] = u;
                        parentEdge[e.to] = i;
                        queue.offer(e.to);
                    }
                }
            }
            if (parent[t] == -1) return flow;

            int bottleneck = Integer.MAX_VALUE;
            for (int v = t; v != s; v = parent[v]) {
                Edge e = graph.get(parent[v]).get(parentEdge[v]);
                bottleneck = Math.min(bottleneck, e.cap);
            }
            for (int v = t; v != s; v = parent[v]) {
                Edge e = graph.get(parent[v]).get(parentEdge[v]);
                e.cap -= bottleneck;
                graph.get(e.to).get(e.rev).cap += bottleneck;
            }
            flow += bottleneck;
        }
    }

    /** BFS the residual graph from s; returns the membership of S side. */
    boolean[] reachableFromSource(int s) {
        boolean[] visited = new boolean[n];
        visited[s] = true;
        Deque<Integer> queue = new ArrayDeque<>();
        queue.offer(s);
        while (!queue.isEmpty()) {
            int u = queue.poll();
            for (Edge e : graph.get(u)) {
                if (e.cap > 0 && !visited[e.to]) {
                    visited[e.to] = true;
                    queue.offer(e.to);
                }
            }
        }
        return visited;
    }

    /** Returns total flow value and the explicit list of cut edges. */
    Object[] minCutEdges(int s, int t) {
        int flow = maxFlow(s, t);
        boolean[] side = reachableFromSource(s);
        List<CutEdge> cut = new ArrayList<>();
        for (int u = 0; u < n; u++) {
            if (!side[u]) continue;
            for (Edge e : graph.get(u)) {
                if (e.original && !side[e.to]) cut.add(new CutEdge(u, e.to, e.originalCap));
            }
        }
        return new Object[]{flow, cut};
    }

    public static void main(String[] args) {
        MinCut g = new MinCut(6);
        int[][] edges = {
            {0, 1, 16}, {0, 2, 13},
            {1, 2, 4},  {1, 3, 12},
            {2, 1, 10}, {2, 4, 14},
            {3, 2, 9},  {3, 5, 20},
            {4, 3, 7},  {4, 5, 4}
        };
        for (int[] e : edges) g.addEdge(e[0], e[1], e[2]);
        Object[] result = g.minCutEdges(0, 5);
        int flow = (int) result[0];
        @SuppressWarnings("unchecked")
        List<CutEdge> cut = (List<CutEdge>) result[1];
        System.out.println("Max flow / Min cut value: " + flow); // 23
        System.out.println("Cut edges (S -> T):");
        for (CutEdge ce : cut) System.out.println("  " + ce);
    }
}

/*
 * NOTES
 * -----
 * Differences from the Python implementation in min_cut.py:
 *   - Java needs explicit Edge / CutEdge classes for tagged residual edges.
 *   - We return an Object[] {flow, List<CutEdge>} since Java lacks anonymous
 *     tuples; an alternative is a small record once on JDK 14+.
 *   - The original capacity is stored on the edge so we can still report it
 *     after the residual has been mutated by max-flow.
 */
